pkg/operator/controller/dns_status.go

```diff
@@ -0,0 +1,173 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// syncDNSStatus computes the current status of dns and
+// updates status upon any changes since last sync.
+func (r *reconciler) syncDNSStatus(dns *operatorv1.DNS, clusterIP, clusterDomain string, ds *appsv1.DaemonSet) error {
+	updated := dns.DeepCopy()
+	updated.Status.ClusterIP = clusterIP
+	updated.Status.ClusterDomain = clusterDomain
+	updated.Status.Conditions = computeDNSStatusConditions(dns.Status.Conditions, clusterIP, ds)
+	if !dnsStatusesEqual(updated.Status, dns.Status) {
+		if err := r.client.Status().Update(context.TODO(), updated); err != nil {
+			return fmt.Errorf("failed to update dns status: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// computeDNSStatusConditions computes dns status conditions
+// based on the status of ds and clusterIP.
+func computeDNSStatusConditions(conditions []operatorv1.OperatorCondition, clusterIP string, ds *appsv1.DaemonSet) []operatorv1.OperatorCondition {
+	conditions = setDNSStatusCondition(conditions, computeDNSDegradedCondition(clusterIP, ds))
+	conditions = setDNSStatusCondition(conditions, computeDNSProgressingCondition(ds))
+	conditions = setDNSStatusCondition(conditions, computeDNSAvailableCondition(clusterIP, ds))
```

Comment:
We could get rid of `setDNSStatusCondition`. Instead of looping all conditions for every `computeDNSXYZCondition` to check for changes, we could compute actual condition once and can patch the LastTransitionTime. Also did not like the fact that LastTransitionTime for the condition is set outside of `computeDNSXYZCondition` method.
WDYT of  https://github.com/openshift/cluster-ingress-operator/pull/200/files#diff-31d813e7a1281abc26f3e03bc7686530R163 ?